security/identity-and-access-management/oracle-identity-governance/scheduled-task-samples/access-extension-notification/files/src/com/oracle/sample/AccessExtensionNotification.java

```diff
@@ -0,0 +1,526 @@
+				sendEmail = processAccounts(userKey, userAccounts);
+
+				if (Boolean.TRUE.equals(sendEmail)) {
+					// retrieve email parameters from the scheduled task
+					String emailTemplateName = (String) arg0.get("Email Template Name");
+					String smtpServerHost = (String) arg0.get("SMTP Mail Server Hostname");
+					Boolean smtpServerTLS = (Boolean) arg0.get("SMTP Mail Server TLS");
+					String smtpServerPort = (String) arg0.get("SMTP Mail Server Port");
+
+					// setup the SMTP mailing server parameters
+					Properties props = new Properties();
+					props.put("mail.smtp.host", smtpServerHost);
+					props.put("mail.transport.protocol", "smtp");
+					props.put("mail.smtp.starttls.enable", smtpServerTLS.toString()); // "true" or "false"
+					props.put("mail.smtp.port", smtpServerPort);
+					Session session = Session.getDefaultInstance(props);
+
+					logger.log(Level.INFO, "Constructing notification email...");
+
+					templateParams = new HashMap<>();
+					templateParams.put("$display_name", currentUser.getDisplayName());
+					templateParams.put("$days", daysBeforeExpiration.toString());
+					templateParams.put("$expiring_items", extensionItemsText.toString());
+					HashMap<String, String> subjectAndBody = dynamicTemplateFill(templateParams, emailTemplateName);
+					emailSubject = subjectAndBody.get("subject");
+					logger.log(Level.FINE, "Filled in email subject: {0}", emailSubject);
+					emailBody = subjectAndBody.get("body");
+					logger.log(Level.FINE, "Filled in email body: {0}", emailBody);
+
+					terminationEmailAddress = currentUser.getEmail();
+
+					if (terminationEmailAddress == null
+							|| (terminationEmailAddress != null && terminationEmailAddress.equals(""))) {
+						throw new MessagingException("The user doesn't have valid email address.");
+					}
+
+					logger.log(Level.FINE, "Send to email address: {0}", terminationEmailAddress);
+
+					logger.log(Level.INFO, "Sending notification email...");
+
+					InternetAddress fromAddress = new InternetAddress(FROM_EMAIL_ADDRESS);
+					InternetAddress toAddress = new InternetAddress(terminationEmailAddress);
+
+					Message message = new MimeMessage(session);
+					message.setFrom(fromAddress);
+					message.setRecipient(Message.RecipientType.TO, toAddress);
+					message.setSubject(emailSubject);
+					message.setContent(emailBody, "text/html; charset=utf-8");
+					message.saveChanges();
+					Transport.send(message);
+
+					logger.log(Level.INFO, "Notification email was sent successfully.");
+				}
+			} catch (MessagingException e) {
+				logger.log(Level.WARNING, "Sending the email failed with: ", e);
+			} catch (Exception e) {
+				logger.log(Level.WARNING, "User processing failed with: ", e);
+			}
+		}
+
+		logger.log(Level.INFO, "The scheduled task run has finished.");
+	}
+
+	private Boolean processAccounts(String userKey, List<Account> userApplications) throws SQLException {
```

Comment:
Can the parameters be empty?

Reply:
Not with the current implementation, however I've added validations in case someone decides to further customize the code.